tensorflow/compiler/tf2tensorrt/convert/ops/quantization_ops.cc

```diff
@@ -0,0 +1,414 @@
+  // while keeping 0 unchanged.
+  const T scale_from_min_side = (min_quantized * *min_range > 0)
+                                    ? min_quantized / *min_range
+                                    : std::numeric_limits<T>::max();
+  const T scale_from_max_side = (max_quantized * *max_range > 0)
+                                    ? max_quantized / *max_range
+                                    : std::numeric_limits<T>::max();
+
+  QuantizationScales<T, 1> scales;
+  // Note: Avoids changing the side of the range that determines scale.
+  if (scale_from_min_side < scale_from_max_side) {
+    scales.quantize_scale[0] = scale_from_min_side;
+    scales.dequantize_scale[0] = *min_range / min_quantized;
+    *max_range = max_quantized * scales.dequantize_scale[0];
+  } else {
+    scales.quantize_scale[0] = scale_from_max_side;
+    scales.dequantize_scale[0] = *max_range / max_quantized;
+    *min_range = min_quantized * scales.dequantize_scale[0];
+  }
+  return scales;
+}
+
+// Prepares the input for a QDQ node in explicit precision mode, returning a
+// ITensor pointer. If the input is weights, we convert to a ITensor by adding a
+// constant layer.
+StatusOr<nvinfer1::ITensor*> ExlicitQDQInputToTensor(
+    TRTNetworkBuilder* builder, OpConverterParams* params,
+    const TRT_TensorOrWeights& input) {
+  if (input.is_tensor()) {
+    return input.tensor()->trt_tensor();
+  }
+  if (!IS_TRT_VERSION_GE(8, 0, 0, 0) && input.weights().count() > 1) {
+    LOG(WARNING) << absl::StrCat(
+        "QDQ per-channel for weights not "
+        "implemented, assuming uniform scaling");
+  }
+  TRT_ShapedWeights trt_weights = input.weights();
+  StatusOr<nvinfer1::IConstantLayer*> weights_const =
+      builder->WeightsToConstant(trt_weights.GetTrtWeights(),
+                                 trt_weights.shape_);
+  TRT_ENSURE_PTR_OK(weights_const);
+  params->converter->SetLayerName(*weights_const, params->node_def, "const");
+  nvinfer1::ITensor* qdq_input = (*weights_const)->getOutput(0);
+  std::string name = absl::StrCat((*weights_const)->getName(), "_output");
+  qdq_input->setName(name.c_str());
+  return qdq_input;
+}
+
+}  // namespace
+
+// Carries traits for each specific quantization op type for conversion.
+// Specialization for template parameter T should be given for each TF C++
+// quantization op.
+template <typename T>
+struct QDQOpSpec {};
+
+template <>
+struct QDQOpSpec<ops::QuantizeAndDequantizeV2> {
+  static constexpr std::array<InputArgSpec, 3> InputSpec() {
+    return {
+        InputArgSpec::Create("input", TrtInputArg::kBoth),
+        InputArgSpec::Create("input_min", TrtInputArg::kWeight),
+        InputArgSpec::Create("input_max", TrtInputArg::kWeight),
+    };
+  }
+
+  struct Attrs {
+    float min_range;
+    float max_range;
+    bool narrow_range;
+    std::string round_mode;
+    UniformQuantizationScales scales;
+  };
+
+  static Status ValidateExplicit(const std::vector<TRT_TensorOrWeights>& inputs,
```

Comment:
It will be clearer to me what this function does if we name it "ValidataQDQForExplicitPrecision"